solver/src/solver/paraswap_solver/api.rs

```diff
@@ -29,45 +29,70 @@ pub struct DefaultParaswapApi {
 
 #[async_trait::async_trait]
 impl ParaswapApi for DefaultParaswapApi {
-    async fn price(&self, query: PriceQuery) -> Result<PriceResponse> {
+    async fn price(&self, query: PriceQuery) -> Result<PriceResponse, ParaswapResponseError> {
+        let query_str = format!("{:?}", &query);
         let url = query.into_url();
         tracing::debug!("Querying Paraswap API (price) for url {}", url);
-        let text = reqwest::get(url)
+        let response_text = reqwest::get(url)
             .await
-            .context("PriceQuery failed")?
+            .map_err(ParaswapResponseError::Send)?
             .text()
-            .await?;
-        tracing::debug!("Response from Paraswap API (price): {}", text);
-
-        serde_json::from_str::<PriceResponse>(&text)
-            .context(format!("PriceQuery result parsing failed: {}", text))
+            .await
+            .map_err(ParaswapResponseError::TextFetch)?;
+        tracing::debug!("Response from Paraswap API (price): {}", response_text);
+        let raw_response = serde_json::from_str::<RawPriceResponse>(&response_text);
+        match raw_response {
+            Ok(RawPriceResponse::ResponseOk(response)) => Ok(response),
+            Ok(RawPriceResponse::ResponseErr { error: message }) => match &message[..] {
+                "computePrice Error" => Err(ParaswapResponseError::ComputePrice(
+                    query_str.parse().unwrap(),
+                )),
+                _ => Err(ParaswapResponseError::UnknownParaswapError(format!(
+                    "uncatalogued Price Query error message {}",
+                    message
+                ))),
+            },
+            Err(err) => Err(ParaswapResponseError::DeserializeError(err)),
+        }
     }
     async fn transaction(
         &self,
         query: TransactionBuilderQuery,
     ) -> Result<TransactionBuilderResponse, ParaswapResponseError> {
         let query_str = serde_json::to_string(&query).unwrap();
-        let response_result = query.into_request(&self.client).send().await;
-        match response_result {
-            Ok(response) => match response.text().await {
-                Ok(response_text) => parse_paraswap_response_text(&response_text, &query_str),
-                Err(text_fetch_err) => Err(ParaswapResponseError::TextFetch(text_fetch_err)),
-            },
-            Err(send_err) => Err(ParaswapResponseError::Send(send_err)),
-        }
+        let response_text = query
+            .into_request(&self.client)
+            .send()
+            .await
+            .map_err(ParaswapResponseError::Send)?
+            .text()
+            .await
+            .map_err(ParaswapResponseError::TextFetch)?;
+        parse_paraswap_response_text(&response_text, &query_str)
     }
 }
 
 #[derive(Deserialize)]
 #[serde(untagged)]
-enum RawResponse {
+enum RawPriceResponse {
```

Comment:
nit: We could make a single `RawResponse` type that can be used by all routes:
```
enum RawResponse<Ok> {
    ResponseOk(Ok),
    ResponseErr { error: String },
}
```

Then, the price route would use `RawResponse<PriceResonse>` and the transaction route would use `RawResponse<TransactionBuilderResponse>`.